### PR TITLE
Paramedic & Security Crew Monitors

### PR DESCRIPTION
**Pull request by rosieposieeee**
## About the PR

Added Paramedic and Security crew monitors, renaming the base one to CMO's. The Paramedic one only shows crit/injured crew, and the Security one only shows Security and Command, crit or not.

Paramedics start with the paramedic one, and Wardens start with the sec one. CMO and HoS have an extra in their lockers as well.

https://github.com/space-wizards/space-station-14/assets/52761126/ac803b38-c556-481f-ad48-6d771c8e2d39

## Why / Balance

Removing the paramedic monitor was understandable, but unfortunate, and I've always wanted a paramedic monitor that _only_ shows crit/dead people for convenience anyway. I can remove the sec monitor from spawning with HoS/Warden if people think its bullshit; it's more there for brigmedic downstreams (of which I play on!) and just as an example of how to use the department/role controls, but having a monitor for prisoners might be really useful.

## Technical details

I modified the crew monitor to take some extra parameters, allowing monitors to select who they want to see. Theoretically you could probably also use this to search in the UI of the full monitor (ie narrow down to just security), but I haven't touched UI stuff yet.

I messed around with adding a noise if someone's crit/dead, but it didn't work out so well, I'd have to learn more about the codebase to get something like that to work. (Needs a mute toggle. By god it needs a mute toggle.)

## Media

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**
:cl:
- add: Added Paramedic and Security Handheld Crew Monitors with more limited functionality.

### CODE REVIEW

**Comment by JackSage** [comment]
Perhaps sec also shows people with tracking implants?

**Comment by Dutch-VanDerLinde** [comment]
> (

normal crew monitor already does this

**Comment by rosieposieeee** [comment]
> Perhaps sec also shows people with tracking implants?

if that was stored in SuitSensorStatus, i would filter it in right now, but the crew monitoring system is not privy to that data and i don't have time rn to look into it; could be a future PR.

**Comment by Yousifb26** [comment]
Is there anyway to replace server not found with something like ‘no crew are currently dying’. It would be a good quality of life to see when the server is down versus it just saying no server because no one is dying

**Comment by rosieposieeee** [comment]
> Is there anyway to replace server not found with something like ‘no crew are currently dying’. It would be a good quality of life to see when the server is down versus it just saying no server because no one is dying

i changed it to say 'no server found or no listings available', which i figured was vague enough to work. unfortunately, _that_ is set just by having no data sent to the monitor, so the monitor literally does not know whether the server is down or no one is injured. like with the tracking implants, i would have to modify code further away from the functionality of this PR, which mostly relies on data that was already available to the system, so that could be a future PR if its too confusing (or i could just make the text more explicit)

**Comment by JackSage** [comment]
> > Perhaps sec also shows people with tracking implants?
> 
> normal crew monitor already does this

I'm saying for the security monitor to show people with tracking implants.

"The Security one only shows Security and Command, crit or not."

File: Content.Server/Medical/CrewMonitoring/CrewMonitoringConsoleComponent.cs [hunk 1]
```diff
@@ -11,6 +11,24 @@ public sealed partial class CrewMonitoringConsoleComponent : Component
...
+    ///     Whether this console shows conscious crew listings or not. If false, will exclusively track dead or crit crew.
+    /// </summary>
+    [DataField]
+    public bool ShowConsciousListings = true;
+
+    /// <summary>
+    ///     Departments to show on this console. Useful for, IE, a security-only monitor for the brigmedic.
+    /// </summary>
+    [DataField]
+    public List<string> TrackedDepartments { get; set; } = ["All"];
```
Comment (deltanedas) [inline Content.Server/Medical/CrewMonitoring/CrewMonitoringConsoleComponent.cs:24]: ProtoId

File: Content.Server/Medical/CrewMonitoring/CrewMonitoringConsoleComponent.cs [hunk 2]
```diff
@@ -11,6 +11,24 @@ public sealed partial class CrewMonitoringConsoleComponent : Component
...
+
+    /// <summary>
+    ///     Specific jobs to show on this console. Useful for, IE, a prisoner-only monitor for the Warden.
+    /// </summary>
+    [DataField]
+    public List<string> TrackedJobs { get; set; } = [];
```
Comment (deltanedas) [inline Content.Server/Medical/CrewMonitoring/CrewMonitoringConsoleComponent.cs:30]: ProtoId

File: Content.Server/Medical/CrewMonitoring/CrewMonitoringConsoleSystem.cs [hunk 3]
```diff
@@ -67,8 +67,58 @@ private void UpdateUserInterface(EntityUid uid, CrewMonitoringConsoleComponent?
...
+        // Get all sensors info
         var allSensors = component.ConnectedSensors.Values.ToList();
-        _uiSystem.SetUiState(bui, new CrewMonitoringState(allSensors));
+        // The sensors we will be outputting to the monitor
+        var outputSensors = new List<SuitSensorStatus>();
+
+        foreach (var listing in allSensors)
+        {
+            // Check to ensure this listing is in one of our tracked departments or roles
+            var is_valid_role = false;
```
Comment (deltanedas) [inline Content.Server/Medical/CrewMonitoring/CrewMonitoringConsoleSystem.cs:78]: rust naming scheme :trollface:

File: Content.Server/Medical/CrewMonitoring/CrewMonitoringConsoleSystem.cs [hunk 4]
```diff
@@ -67,8 +67,58 @@ private void UpdateUserInterface(EntityUid uid, CrewMonitoringConsoleComponent?
...
+            {
+                // Only add conscious listings if we're supposed to
+                var is_valid_listing = false;
+                if (component.ShowConsciousListings)
+                {
+                    is_valid_listing = true;
+                }
+                else
+                {
+                    if (listing.TotalDamage > 100)
```
Comment (deltanedas) [inline Content.Server/Medical/CrewMonitoring/CrewMonitoringConsoleSystem.cs:108]: use mobstate instead

File: Resources/Prototypes/Entities/Objects/Specific/Medical/handheld_crew_monitor.yml [hunk 5]
```diff
@@ -1,10 +1,9 @@
 - type: entity
+  abstract: true
   name: handheld crew monitor
-  suffix: DO NOT MAP
   parent:
   - BaseItem
   - PowerCellSlotSmallItem
-  # CMO-only bud, don't add more.
   id: HandheldCrewMonitor
```
Comment (deltanedas) [inline Resources/Prototypes/Entities/Objects/Specific/Medical/handheld_crew_monitor.yml:7]: breaking change, make this BaseHandheldCrewMonitor

File: Resources/Prototypes/Entities/Objects/Specific/Medical/handheld_crew_monitor.yml [hunk 6]
```diff
@@ -31,15 +31,71 @@
...
+        name: power-cell-slot-component-slot-name-default
+
+- type: entity
+  name: paramedic handheld crew monitor
   parent: HandheldCrewMonitor
+  id: ParamedicHandheldCrewMonitor
+  description: A paramedic-specific hand-held crew monitor displaying the status of suit sensors. It only shows crew in a critical state or already dead.
+  components:
+  - type: Sprite
+    sprite: Objects/Specific/Medical/handheldcrewmonitor.rsi
```
Comment (deltanedas) [inline Resources/Prototypes/Entities/Objects/Specific/Medical/handheld_crew_monitor.yml:64]: inherited so remove this line

File: Resources/Prototypes/Entities/Objects/Specific/Medical/handheld_crew_monitor.yml [hunk 7]
```diff
@@ -31,15 +31,71 @@
...
+    state: security
+  - type: CrewMonitoringConsole
+    trackedDepartments:
+      - Command
+      - Security
+    trackedJobs:
+      - Prisoner
```
Comment (deltanedas) [inline Resources/Prototypes/Entities/Objects/Specific/Medical/handheld_crew_monitor.yml:94]: :trollface: if you used ProtoId youd have already caught this one

**Review by deltanedas — CHANGES_REQUESTED** [review]

File: Content.Server/Medical/CrewMonitoring/CrewMonitoringConsoleSystem.cs (hunk 3 above)
Comment (rosieposieeee) [reply Content.Server/Medical/CrewMonitoring/CrewMonitoringConsoleSystem.cs:78]: orz

File: Resources/Prototypes/Entities/Objects/Specific/Medical/handheld_crew_monitor.yml (hunk 6 above)
Comment (rosieposieeee) [reply Resources/Prototypes/Entities/Objects/Specific/Medical/handheld_crew_monitor.yml:64]: the sprite changed and the description is different...?

File: Resources/Prototypes/Entities/Objects/Specific/Medical/handheld_crew_monitor.yml (hunk 7 above)
Comment (rosieposieeee) [reply Resources/Prototypes/Entities/Objects/Specific/Medical/handheld_crew_monitor.yml:94]: image

it works, no?

File: Content.Server/Medical/CrewMonitoring/CrewMonitoringConsoleSystem.cs (hunk 4 above)
Comment (rosieposieeee) [reply Content.Server/Medical/CrewMonitoring/CrewMonitoringConsoleSystem.cs]: did this but for some reason this.. just... doesn't work...???

image

**Comment by Simyon264** [comment]
I like the paramedic crew monitor but why would security need one?

**Comment by blueDev2** [comment]
> I like the paramedic crew monitor but why would security need one?

In some ss13 forks, one of the warden's jobs is to check the health of the officers. It's useful for sec to know if and where their officers' dead bodies are. Might have some balance implications though.

**Comment by rosieposieeee** [comment]
> I like the paramedic crew monitor but why would security need one?

crazily enough there is a Why / Balance section of the pr. that explains this

**Comment by rosieposieeee** [comment]
> looks like you forgot to add to migration yml

ah, more like didn't know this was a thing, whoops. guess i'll remove that map change

**Comment by rosieposieeee** [comment]
i can remove the sec one if thats the main holdup for merging this

**Comment by JackSage** [comment]
I thought a cool way to do this instead would be for paramedics to get a special pinpointer (not the nuke pinpointer), by default it would track nothing. The paramedic can go to a crew monitoring console and they can select the pinpointer to track someone. (i.e. if John Doe is crit w/ cords on, you use the crew monitor, track him, and then the pin pointer will point to him) The pinpointer will stop tracking if they turn their cords off, a jammer is active near them, or their jumpsuit/skirt is removed.

File: Resources/Prototypes/Entities/Objects/Specific/Medical/handheld_crew_monitor.yml (hunk 6 above)
Comment (lzk228) [reply Resources/Prototypes/Entities/Objects/Specific/Medical/handheld_crew_monitor.yml:72]: the parent `BaseHandheldCrewMonitor` already has the path written so you need just to change state for children

File: Resources/Prototypes/Entities/Objects/Specific/Medical/handheld_crew_monitor.yml [hunk 8]
```diff
@@ -31,15 +39,77 @@
...
+        name: power-cell-slot-component-slot-name-default
+
+- type: entity
+  name: security handheld crew monitor
+  parent: BaseHandheldCrewMonitor
+  id: SecurityHandheldCrewMonitor
+  description: A security-specific hand-held crew monitor displaying the status of suit sensors.
+  components:
+  - type: Sprite
+    sprite: Objects/Specific/Medical/handheldcrewmonitor.rsi
```
Comment (lzk228) [inline Resources/Prototypes/Entities/Objects/Specific/Medical/handheld_crew_monitor.yml:103]: same here

**Review by lzk228 — CHANGES_REQUESTED** [review]
in everything else yml part looks fine to me

**Comment by LankLTE** [comment]
Sec having their own crew monitor just to bypass med isn't something we want. Paramedic having their own crew monitor makes the CMO's less unique as an item and kind of comes across like a way to revert the change.

**Comment by rosieposieeee** [comment]
> Sec having their own crew monitor just to bypass med isn't something we want. Paramedic having their own crew monitor makes the CMO's less unique as an item and kind of comes across like a way to revert the change.

what about the code changes i did that allow them to exist, for downstreams? i know you guys 'dont do things for downstreams' but its impossible to do this with yml changes alone.

**Pull request closed without merging**